### PR TITLE
improvement(availability zone): Fallback availability zones in AWS instance provisioning

### PR DESCRIPTION
**Pull request by yarongilor**
In order to avoid a specific AWS availability zone instance capacity limitation,
        SCT can try all availability zones one by one in order to maximize 
        the chances of getting the requested instance capacity.
Fixes: https://github.com/scylladb/scylla-cluster-tests/issues/5119
https://trello.com/c/5rg6tNqD
## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [x] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by yarongilor** [comment]
Running a test in: https://jenkins.scylladb.com/job/scylla-staging/job/Longevity_yaron/job/artifacts-ami-arm-test-az/12/

Have to run some additional tests to see if this approach is good enough or it would still be better to add an explicit availability-zone and a retry mechanizm.

File: sdcm/utils/aws_utils.py [hunk 1]
```diff
@@ -263,13 +265,20 @@ def get_common_params(params: dict, regions: List, credentials: List, services:
                 )
 
 
-def get_ec2_network_configuration(regions: list[str], availability_zones: list[str], params: dict):
+def get_ec2_network_configuration(regions: list[str], availability_zones: list[str], params: dict,
+                                  auto_availability_zone: bool = False):
```
Comment (fruch) [inline sdcm/utils/aws_utils.py:269]: why `auto_availability_zone` is needed to be added all the way down, if params is being passed to the functions

just do the params.get('auto_availability_zone') down in this function

I think there are more places  `get_common_params` or `get_ec2_network_configuration` get called

File: sdcm/sct_config.py [hunk 2]
```diff
@@ -1222,6 +1222,10 @@ class SCTConfiguration(dict):
         dict(name="availability_zone", env="SCT_AVAILABILITY_ZONE",
              type=str,
              help="Availability zone to use. Same for multi-region scenario."),
+        dict(name="auto_availability_zone", env="AUTO_AVAILABILITY_ZONE",
```
Comment (fruch) [inline sdcm/sct_config.py:1225]: the environment variable should be prefixed with `SCT_`

also it AWS only feature, so it's config option should be prefixed with `aws`

File: sdcm/utils/aws_utils.py (hunk 1 above)
Comment (yarongilor) [reply sdcm/utils/aws_utils.py:269]: Yes, there is one more place i didn't add it intentionally since it is in scope of k8s. No problem adding it there as well. @vponomaryov can you please advise - can there be any future need for this enhancement in k8s?

Comment (fruch) [reply sdcm/utils/aws_utils.py:269]: it's a feature related to the AWS backend, k8s probably doesn't needed it now (same as longevity doesn't needed it)

This feature is currently only for artifacts tests, and there no real need of changing k8s code.

i.e. you can encapsulate the change only to `get_ec2_network_configuration`, no change those function signatures

File: sdcm/sct_config.py (hunk 2 above)
Comment (fruch) [reply sdcm/sct_config.py:1225]: Also add a default `False` in `test_default.yaml` for this configuration

File: sdcm/utils/aws_utils.py (hunk 1 above)
Comment (yarongilor) [reply sdcm/utils/aws_utils.py]: fixed

File: sdcm/sct_config.py (hunk 2 above)
Comment (yarongilor) [reply sdcm/sct_config.py]: fixed

File: sdcm/utils/aws_utils.py [hunk 3]
```diff
@@ -268,8 +268,14 @@ def get_ec2_network_configuration(regions: list[str], availability_zones: list[s
...
         aws_region = AwsRegion(region_name=region)
+        if params.get('auto_availability_zone'):
+            availability_zones = aws_region.availability_zones
+        else:
+            availability_zones = [region + availability_zone for availability_zone in availability_zones]
+
+        LOGGER.debug("Getting subnets and SGs for availability zones of: %s", availability_zones)
         for availability_zone in availability_zones:
-            sct_subnet = aws_region.sct_subnet(region_az=region + availability_zone)
+            sct_subnet = aws_region.sct_subnet(region_az=availability_zone)
```
Comment (roydahan) [inline sdcm/utils/aws_utils.py:278]: I'm sorry, but I don't understand how it works or if it works as the requirement in the task.
(And if there is such a magic it should be explained in the commit message).

The requirement is to catch error of AZ that can't fullfill the provision and retry it on the other AZs.
What you did here is not specifying AZ?
Is there any promise that AWS will be able to provision the instances like this?
How is it even possible here that you first get the sunbet of the AZ even before you try to provision...

File: sdcm/sct_config.py [hunk 4]
```diff
@@ -1222,6 +1222,10 @@ class SCTConfiguration(dict):
         dict(name="availability_zone", env="SCT_AVAILABILITY_ZONE",
              type=str,
              help="Availability zone to use. Same for multi-region scenario."),
+        dict(name="auto_availability_zone", env="SCT_AWS_AUTO_AVAILABILITY_ZONE",
```
Comment (vponomaryov) [inline sdcm/sct_config.py:1225]: Having env var be called `SCT_AWS_AUTO_AVAILABILITY_ZONE` the python var must be named as `aws_auto_availability_zone`.

File: sdcm/sct_config.py [hunk 5]
```diff
@@ -1222,6 +1222,10 @@ class SCTConfiguration(dict):
...
+             type=boolean,
+             help="Not limiting availability zones, in order for the cloud provider to automatically select"
```
Comment (vponomaryov) [inline sdcm/sct_config.py:1227]: Since it is backend-specific the description must mention it.
But, I guess @soyacz will have concerns because of the provision logic unification process violation.

Comment (yarongilor) [reply sdcm/sct_config.py:1227]: right, that's the cocern i had as well, that although only implement currently for AWS, it can be similarly added in the future to other providers.

Comment (fgelcer) [reply sdcm/sct_config.py:1227]: @soyacz , could you please review this PR?

Comment (soyacz) [reply sdcm/sct_config.py:1227]: can we just not set availability_zone and consider automatic if not set instead of adding new variable?

**Review by soyacz — CHANGES_REQUESTED** [review]
I also don't understand how modification to `get_ec2_network_configuration` solves the problem stated. Can you elaborate solution how it works?

File: sdcm/sct_config.py (hunk 5 above)
Comment (roydahan) [reply sdcm/sct_config.py:1227]: This whole automatic is not the requirement and not part of the task.

Comment (yarongilor) [reply sdcm/sct_config.py:1227]: @roydahan , @fgelcer , there can be 2 different ways for a solution:
1) Catch 'no resources' and retry another zone. (As stated in AWS error message)
2) Not sending AZ related requirements to AWS (as recommended in AWS error message as well). This is what asked in issue 5519 and by Lukasz above.
So we have to decide one of the above.

Comment (roydahan) [reply sdcm/sct_config.py:1227]: The solution is stated in the task description.
We considered other solutions and don't think they're a good fit.

**Comment by yarongilor** [comment]
Latest test seems to run ok and fail as expected after exhauted all availability zones, encountering capacity errors:
( https://jenkins.scylladb.com/job/scylla-staging/job/Longevity_yaron/job/artifacts-ami-arm-test-az/20/ )
```
09:15:41  < t:2022-08-29 06:15:40,830 f:cluster_aws.py  l:372  c:sdcm.cluster         p:INFO  > Cluster artifacts-ami-jenkins-db-cluster-66b1c933 (AMI: ['ami-0f3c56bf20c33ce6b'] Type: is4gen.8xlarge): Found no provisioned instances. Provision them.
09:15:41  < t:2022-08-29 06:15:40,831 f:cluster_aws.py  l:213  c:sdcm.cluster         p:INFO  > Cluster artifacts-ami-jenkins-db-cluster-66b1c933 (AMI: ['ami-0f3c56bf20c33ce6b'] Type: is4gen.8xlarge): Create on_demand instance(s)
09:15:51  < t:2022-08-29 06:15:51,305 f:tester.py       l:1093 c:ArtifactsTest        p:WARNING > Failed creating a Scylla AWS cluster: An error occurred (InsufficientInstanceCapacity) when calling the RunInstances operation (reached max retries: 4): We currently do not have sufficient is4gen.8xlarge capacity in the Availability Zone you requested (us-east-1a). Our system will be working on provisioning additional capacity. You can currently get is4gen.8xlarge capacity by not specifying an Availability Zone in your request or choosing us-east-1b, us-east-1c, us-east-1d.
09:15:51  < t:2022-08-29 06:15:51,307 f:cluster.py      l:2977 c:sdcm.cluster         p:INFO  > Cluster artifacts-ami-jenkins-db-cluster-66b1c933 (AMI: ['ami-0f3c56bf20c33ce6b'] Type: is4gen.8xlarge): Init nodes
09:15:52  < t:2022-08-29 06:15:52,296 f:cluster_aws.py  l:372  c:sdcm.cluster         p:INFO  > Cluster artifacts-ami-jenkins-db-cluster-66b1c933 (AMI: ['ami-0f3c56bf20c33ce6b'] Type: is4gen.8xlarge): Found no provisioned instances. Provision them.
09:15:52  < t:2022-08-29 06:15:52,297 f:cluster_aws.py  l:213  c:sdcm.cluster         p:INFO  > Cluster artifacts-ami-jenkins-db-cluster-66b1c933 (AMI: ['ami-0f3c56bf20c33ce6b'] Type: is4gen.8xlarge): Create on_demand instance(s)
09:16:05  < t:2022-08-29 06:16:04,229 f:tester.py       l:1093 c:ArtifactsTest        p:WARNING > Failed creating a Scylla AWS cluster: An error occurred (InsufficientInstanceCapacity) when calling the RunInstances operation (reached max retries: 4): We currently do not have sufficient is4gen.8xlarge capacity in the Availability Zone you requested (us-east-1b). Our system will be working on provisioning additional capacity. You can currently get is4gen.8xlarge capacity by not specifying an Availability Zone in your request or choosing us-east-1a, us-east-1c, us-east-1d.
09:16:05  < t:2022-08-29 06:16:04,231 f:cluster.py      l:2977 c:sdcm.cluster         p:INFO  > Cluster artifacts-ami-jenkins-db-cluster-66b1c933 (AMI: ['ami-0f3c56bf20c33ce6b'] Type: is4gen.8xlarge): Init nodes
09:16:05  < t:2022-08-29 06:16:04,692 f:cluster_aws.py  l:372  c:sdcm.cluster         p:INFO  > Cluster artifacts-ami-jenkins-db-cluster-66b1c933 (AMI: ['ami-0f3c56bf20c33ce6b'] Type: is4gen.8xlarge): Found no provisioned instances. Provision them.
09:16:05  < t:2022-08-29 06:16:04,694 f:cluster_aws.py  l:213  c:sdcm.cluster         p:INFO  > Cluster artifacts-ami-jenkins-db-cluster-66b1c933 (AMI: ['ami-0f3c56bf20c33ce6b'] Type: is4gen.8xlarge): Create on_demand instance(s)
09:16:14  < t:2022-08-29 06:16:13,579 f:tester.py       l:1093 c:ArtifactsTest        p:WARNING > Failed creating a Scylla AWS cluster: An error occurred (InsufficientInstanceCapacity) when calling the RunInstances operation (reached max retries: 4): We currently do not have sufficient is4gen.8xlarge capacity in the Availability Zone you requested (us-east-1c). Our system will be working on provisioning additional capacity. You can currently get is4gen.8xlarge capacity by not specifying an Availability Zone in your request or choosing us-east-1a, us-east-1b, us-east-1d.
09:16:14  < t:2022-08-29 06:16:13,580 f:cluster.py      l:2977 c:sdcm.cluster         p:INFO  > Cluster artifacts-ami-jenkins-db-cluster-66b1c933 (AMI: ['ami-0f3c56bf20c33ce6b'] Type: is4gen.8xlarge): Init nodes
09:16:14  < t:2022-08-29 06:16:14,335 f:cluster_aws.py  l:372  c:sdcm.cluster         p:INFO  > Cluster artifacts-ami-jenkins-db-cluster-66b1c933 (AMI: ['ami-0f3c56bf20c33ce6b'] Type: is4gen.8xlarge): Found no provisioned instances. Provision them.
09:16:14  < t:2022-08-29 06:16:14,335 f:cluster_aws.py  l:213  c:sdcm.cluster         p:INFO  > Cluster artifacts-ami-jenkins-db-cluster-66b1c933 (AMI: ['ami-0f3c56bf20c33ce6b'] Type: is4gen.8xlarge): Create on_demand instance(s)
09:16:25  < t:2022-08-29 06:16:24,255 f:tester.py       l:1093 c:ArtifactsTest        p:WARNING > Failed creating a Scylla AWS cluster: An error occurred (InsufficientInstanceCapacity) when calling the RunInstances operation (reached max retries: 4): We currently do not have sufficient is4gen.8xlarge capacity in the Availability Zone you requested (us-east-1d). Our system will be working on provisioning additional capacity. You can currently get is4gen.8xlarge capacity by not specifying an Availability Zone in your request or choosing us-east-1a, us-east-1b, us-east-1c.
09:16:25  < t:2022-08-29 06:16:24,258 f:cluster.py      l:2977 c:sdcm.cluster         p:INFO  > Cluster artifacts-ami-jenkins-db-cluster-66b1c933 (AMI: ['ami-0f3c56bf20c33ce6b'] Type: is4gen.8xlarge): Init nodes
09:16:25  < t:2022-08-29 06:16:25,114 f:cluster_aws.py  l:372  c:sdcm.cluster         p:INFO  > Cluster artifacts-ami-jenkins-db-cluster-66b1c933 (AMI: ['ami-0f3c56bf20c33ce6b'] Type: is4gen.8xlarge): Found no provisioned instances. Provision them.
09:16:25  < t:2022-08-29 06:16:25,115 f:cluster_aws.py  l:213  c:sdcm.cluster         p:INFO  > Cluster artifacts-ami-jenkins-db-cluster-66b1c933 (AMI: ['ami-0f3c56bf20c33ce6b'] Type: is4gen.8xlarge): Create on_demand instance(s)
09:16:26  < t:2022-08-29 06:16:25,701 f:tester.py       l:1093 c:ArtifactsTest        p:WARNING > Failed creating a Scylla AWS cluster: An error occurred (Unsupported) when calling the RunInstances operation: Your requested instance type (is4gen.8xlarge) is not supported in your requested Availability Zone (us-east-1e). Please retry your request by not specifying an Availability Zone or choosing us-east-1a, us-east-1b, us-east-1c, us-east-1d.
09:16:26  < t:2022-08-29 06:16:25,702 f:cluster.py      l:2977 c:sdcm.cluster         p:INFO  > Cluster artifacts-ami-jenkins-db-cluster-66b1c933 (AMI: ['ami-0f3c56bf20c33ce6b'] Type: is4gen.8xlarge): Init nodes
09:16:26  < t:2022-08-29 06:16:26,387 f:cluster_aws.py  l:372  c:sdcm.cluster         p:INFO  > Cluster artifacts-ami-jenkins-db-cluster-66b1c933 (AMI: ['ami-0f3c56bf20c33ce6b'] Type: is4gen.8xlarge): Found no provisioned instances. Provision them.
09:16:26  < t:2022-08-29 06:16:26,388 f:cluster_aws.py  l:213  c:sdcm.cluster         p:INFO  > Cluster artifacts-ami-jenkins-db-cluster-66b1c933 (AMI: ['ami-0f3c56bf20c33ce6b'] Type: is4gen.8xlarge): Create on_demand instance(s)
09:16:27  E
09:16:27  ======================================================================
09:16:27  ERROR: test_scylla_service (artifacts_test.ArtifactsTest)
09:16:27  ----------------------------------------------------------------------
09:16:27  Traceback (most recent call last):
09:16:27    File "/home/jenkins/slave/workspace/scylla-staging/Longevity_yaron/artifacts-ami-arm-test-az/scylla-cluster-tests/artifacts_test.py", line 44, in setUp
09:16:27      super().setUp()
09:16:27    File "/home/jenkins/slave/workspace/scylla-staging/Longevity_yaron/artifacts-ami-arm-test-az/scylla-cluster-tests/sdcm/tester.py", line 146, in wrapper
09:16:27      return method(*args, **kwargs)
09:16:27    File "/home/jenkins/slave/workspace/scylla-staging/Longevity_yaron/artifacts-ami-arm-test-az/scylla-cluster-tests/sdcm/utils/decorators.py", line 116, in inner
09:16:27      res = func(*args, **kwargs)
09:16:27    File "/home/jenkins/slave/workspace/scylla-staging/Longevity_yaron/artifacts-ami-arm-test-az/scylla-cluster-tests/sdcm/tester.py", line 696, in setUp
09:16:27      self.init_resources()
09:16:27    File "/home/jenkins/slave/workspace/scylla-staging/Longevity_yaron/artifacts-ami-arm-test-az/scylla-cluster-tests/sdcm/tester.py", line 1616, in init_resources
09:16:27      self.get_cluster_aws(loader_info=loader_info, db_info=db_info,
09:16:27    File "/home/jenkins/slave/workspace/scylla-staging/Longevity_yaron/artifacts-ami-arm-test-az/scylla-cluster-tests/sdcm/tester.py", line 1149, in get_cluster_aws
09:16:27      self.db_cluster = create_cluster(db_type)
09:16:27    File "/home/jenkins/slave/workspace/scylla-staging/Longevity_yaron/artifacts-ami-arm-test-az/scylla-cluster-tests/sdcm/tester.py", line 1111, in create_cluster
09:16:27      return _create_auto_zone_scylla_aws_cluster()
09:16:27    File "/home/jenkins/slave/workspace/scylla-staging/Longevity_yaron/artifacts-ami-arm-test-az/scylla-cluster-tests/sdcm/tester.py", line 1097, in _create_auto_zone_scylla_aws_cluster
09:16:27      raise CriticalTestFailure("Failed creating a Scylla AWS cluster: %s", str(capacity_errors))
09:16:27  sdcm.tester.CriticalTestFailure: ('Failed creating a Scylla AWS cluster: %s', "[ClientError('An error occurred (InsufficientInstanceCapacity) when calling the RunInstances operation (reached max retries: 4): We currently do not have sufficient is4gen.8xlarge capacity in the Availability Zone you requested (us-east-1a). Our system will be working on provisioning additional capacity. You can currently get is4gen.8xlarge capacity by not specifying an Availability Zone in your request or choosing us-east-1b, us-east-1c, us-east-1d.'), ClientError('An error occurred (InsufficientInstanceCapacity) when calling the RunInstances operation (reached max retries: 4): We currently do not have sufficient is4gen.8xlarge capacity in the Availability Zone you requested (us-east-1b). Our system will be working on provisioning additional capacity. You can currently get is4gen.8xlarge capacity by not specifying an Availability Zone in your request or choosing us-east-1a, us-east-1c, us-east-1d.'), ClientError('An error occurred (InsufficientInstanceCapacity) when calling the RunInstances operation (reached max retries: 4): We currently do not have sufficient is4gen.8xlarge capacity in the Availability Zone you requested (us-east-1c). Our system will be working on provisioning additional capacity. You can currently get is4gen.8xlarge capacity by not specifying an Availability Zone in your request or choosing us-east-1a, us-east-1b, us-east-1d.'), ClientError('An error occurred (InsufficientInstanceCapacity) when calling the RunInstances operation (reached max retries: 4): We currently do not have sufficient is4gen.8xlarge capacity in the Availability Zone you requested (us-east-1d). Our system will be working on provisioning additional capacity. You can currently get is4gen.8xlarge capacity by not specifying an Availability Zone in your request or choosing us-east-1a, us-east-1b, us-east-1c.'), ClientError('An error occurred (Unsupported) when calling the RunInstances operation: Your requested instance type (is4gen.8xlarge) is not supported in your requested Availability Zone (us-east-1e). Please retry your request by not specifying an Availability Zone or choosing us-east-1a, us-east-1b, us-east-1c, us-east-1d.'), ClientError('An error occurred (Unsupported) when calling the RunInstances operation: Your requested instance type (is4gen.8xlarge) is not supported in your requested Availability Zone (us-east-1f). Please retry your request by not specifying an Availability Zone or choosing us-east-1a, us-east-1b, us-east-1c, us-east-1d.')]")
09:16:27  
09:16:27  ----------------------------------------------------------------------
09:16:27  Ran 1 test in 50.580s
09:16:27  
```

**Comment by yarongilor** [comment]
> @yarongilor this change isn't doing what you think, having a list of zone, just translate to having a list of subnet_id, and later used like that in `cluster_aws.py`:
> 
> ```
> _ec2_subnet_id[dc_idx]
> ```
> 
> in artifacts test `dc_idx` would be always zero, and hence you changes seems to be working, selecting the first zone always. it won't help if the instance type isn't available, and we'll never use any of the other zones like that.
> 
> Creating instances we can not select multiple subnets/zones, only one. (only spot fleets and Auto Scaling Groups, support such a thing)
> 
> Please stick to what suggest, of implementing a retry with the next zone in the list, for the cases we fail on instance not available

thanks @fruch , i think this is quite a misleading and incorrect implementation that better deserve a refactor one day. It means sending availability zones list is quite meaningless in this case, for example.
anyway i reimplemeted it according to the original suggestion now and it runs ok.

File: sdcm/sct_config.py (hunk 4 above)
Comment (yarongilor) [reply sdcm/sct_config.py]: fixed

File: sdcm/sct_config.py (hunk 5 above)
Comment (yarongilor) [reply sdcm/sct_config.py:1227]: fixed

**Comment by yarongilor** [comment]
> I also don't understand how modification to `get_ec2_network_configuration` solves the problem stated. Can you elaborate solution how it works?

@soyacz , in current sct aws implementation there is a different subnet and SG in every region availability zone. Thus it is selected according to te specified availability zone and sent as part of the request parameters of AWS EC2 API RunInstance.

**Comment by fruch** [comment]
> > @yarongilor this change isn't doing what you think, having a list of zone, just translate to having a list of subnet_id, and later used like that in `cluster_aws.py`:
> > ```
> > _ec2_subnet_id[dc_idx]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > in artifacts test `dc_idx` would be always zero, and hence you changes seems to be working, selecting the first zone always. it won't help if the instance type isn't available, and we'll never use any of the other zones like that.
> > Creating instances we can not select multiple subnets/zones, only one. (only spot fleets and Auto Scaling Groups, support such a thing)
> > Please stick to what suggest, of implementing a retry with the next zone in the list, for the cases we fail on instance not available
> 
> thanks @fruch , i think this is quite a misleading and incorrect implementation that better deserve a refactor one day. It means sending availability zones list is quite meaningless in this case, for example. anyway i reimplemeted it according to the original suggestion now and it runs ok.

it was built for supporting multi-DCs, not for supporting multiple AZs (anyhow you can't have EC2 instance connect to multiple AZs at the same time), also we don't want different nodes in different AZs, the traffic would be very expansive for us.

File: sdcm/utils/aws_utils.py (hunk 3 above)
Comment (yarongilor) [reply sdcm/utils/aws_utils.py:275]: fixed

File: sdcm/sct_config.py [hunk 6]
```diff
@@ -1222,6 +1222,10 @@ class SCTConfiguration(dict):
         dict(name="availability_zone", env="SCT_AVAILABILITY_ZONE",
              type=str,
              help="Availability zone to use. Same for multi-region scenario."),
+        dict(name="aws_auto_availability_zone", env="SCT_AWS_AUTO_AVAILABILITY_ZONE",
+             type=boolean,
+             help="Not limiting availability zones, in order for the cloud provider to automatically select"
```
Comment (fruch) [inline sdcm/sct_config.py:1227]: please update the description, and rename `aws_auto_availability_zone`, it's a fallback similar to `instance_provision_fallback_on_demand`

so something like `aws_fallback_to_next_availability_zone` might be clearer.

File: sdcm/utils/aws_utils.py [hunk 7]
```diff
@@ -268,8 +269,10 @@ def get_ec2_network_configuration(regions: list[str], availability_zones: list[s
     ec2_subnet_ids = []
     for region in regions:
         aws_region = AwsRegion(region_name=region)
+        availability_zones = [region + availability_zone for availability_zone in availability_zones]
```
Comment (fruch) [inline sdcm/utils/aws_utils.py:272]: can you revert this changes ? they don't seem to do anything. just adding a print.

**Review by fruch — CHANGES_REQUESTED** [review]
Rename the config option, and update the description (also in the commit/PR and headline)

File: sdcm/sct_config.py (hunk 6 above)
Comment (yarongilor) [reply sdcm/sct_config.py]: fixed

**Comment by yarongilor** [comment]
> In order to avoid a specific AWS availability zone instance capacity limitation,
>     SCT can try all availabilit

fixed

File: sdcm/sct_config.py [hunk 8]
```diff
@@ -1222,6 +1222,10 @@ class SCTConfiguration(dict):
         dict(name="availability_zone", env="SCT_AVAILABILITY_ZONE",
              type=str,
              help="Availability zone to use. Same for multi-region scenario."),
+        dict(name="aws_fallback_to_next_availability_zone", env="SCT_AWS_AUTO_AVAILABILITY_ZONE",
```
Comment (fruch) [inline sdcm/sct_config.py:1225]:
```suggestion
        dict(name="aws_fallback_to_next_availability_zone", env="SCT_AWS_FALLBACK_TO_NEXT_AVAILABILITY_ZONE",
```

File: sdcm/utils/aws_utils.py (hunk 7 above)
Comment (fruch) [reply sdcm/utils/aws_utils.py:272]: @yarongilor please revert those unneeded changes

**Review by soyacz — CHANGES_REQUESTED** [review]
Unfortunately I'm not sure if this solution works:
1. you tested with artifact test which is not using provisioner used in separate provision-step used in longevities. Artifact test is not using that part of code.
2. How it will behave when we have sct-runner in one AZ while resources in another? We should also connect runner to resources subnet as well i think

besides that, I did some investigation about solving our problem and found that AWS have an api to recommend best region/AZ for given requested resources: https://docs.aws.amazon.com/cli/latest/reference/ec2/get-spot-placement-scores.html

This could be also solution to this problem and could be much cleaner (e.g. based on test resources, before creating sct runner, get best AZ (or even region) for test and create runner and resources there.)

File: sdcm/sct_config.py (hunk 8 above)
Comment (yarongilor) [reply sdcm/sct_config.py:1225]: fixing

**Comment by yarongilor** [comment]
> re if this solution works:
> 
>     1. you tested with artifact test which is not using provisioner used in separate provision-step used in longevities. Artifact test is not using that part of code.
> 
>     2. How it will behave when we have sct-runner in one AZ while resources in another? We should also connect runner to resources subnet as well i think
> 
> 
> besides that, I did some investigation about solving our problem and found that AWS have an api to recommend best region/AZ for given requested resources: https://docs.aws.amazon.com/cli/latest/reference/ec2/get-spot-placement-scores.html
> 
> This could be also solution to this problem and could be much cleaner (e.g. based on test resources, before creating sct runner, get best AZ (or even region) for test and create runner and resources there.)

@soyacz , 
1. in principle and for the long term it sounds correct. but as you can note, the goal for this task is only a minimal fix to mainly solve the problem we face in artifacts test. So it is not involved with any refactor, just a patch.
2. artifacts test doesn't use sct runner.
3. This fix was tested and passed ok for artifacts test.
4. I'm not sure it is a problem if sct runner located in a different zone.

**Review by soyacz — APPROVED** [review]

**Comment by soyacz** [comment]
> > re if this solution works:
> > ```
> > 1. you tested with artifact test which is not using provisioner used in separate provision-step used in longevities. Artifact test is not using that part of code.
> > 
> > 2. How it will behave when we have sct-runner in one AZ while resources in another? We should also connect runner to resources subnet as well i think
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > besides that, I did some investigation about solving our problem and found that AWS have an api to recommend best region/AZ for given requested resources: https://docs.aws.amazon.com/cli/latest/reference/ec2/get-spot-placement-scores.html
> > This could be also solution to this problem and could be much cleaner (e.g. based on test resources, before creating sct runner, get best AZ (or even region) for test and create runner and resources there.)
> 
> @soyacz ,
> 
> 1. in principle and for the long term it sounds correct. but as you can note, the goal for this task is only a minimal fix to mainly solve the problem we face in artifacts test. So it is not involved with any refactor, just a patch.
> 2. artifacts test doesn't use sct runner.
> 3. This fix was tested and passed ok for artifacts test.
> 4. I'm not sure it is a problem if sct runner located in a different zone.

Ok, in a case we want this only for artifact test where we don't use separate provision step nor sct-runner, it can be accepted.
Anyway, It would be beneficial and not to difficult to implement additional step in pipelines to select best place for test @roydahan  @fgelcer what do you think?

**Comment by fruch** [comment]
> Unfortunately I'm not sure if this solution works:
> 1. you tested with artifact test which is not using provisioner used in separate provision-step used in longevities. Artifact test is not using that part of code.

That acceptable, if we'll need it also in provision code, we'll implement there as well (it's less needed there since we don't want mixed AZ in a test, the traffic cost is too high)

> 2. How it will behave when we have sct-runner in one AZ while resources in another? We should also connect runner to resources subnet as well i think

Again it's only artifacts test, and most of the traffic is between nodes and loaders, but also logs into the runner, we should strive to make sure we run in the same subnet. so not much point for this when SCT runner is used.

> 
> besides that, I did some investigation about solving our problem and found that AWS have an api to recommend best region/AZ for given requested resources: https://docs.aws.amazon.com/cli/latest/reference/ec2/get-spot-placement-scores.html
> 
> This could be also solution to this problem and could be much cleaner (e.g. based on test resources, before creating sct runner, get best AZ (or even region) for test and create runner and resources there.)

Using this feature could be nice, but not necessarily simpler, it would be nice to check all the instances up front and automatically pick regions and availability zone, or based on pricing. but it's out of scope for this task.

**Review by fruch — APPROVED** [review]
LGTM

**Comment by fgelcer** [comment]
> Ok, in a case we want this only for artifact test where we don't use separate provision step nor sct-runner, it can be accepted. Anyway, It would be beneficial and not to difficult to implement additional step in pipelines to select best place for test @roydahan @fgelcer what do you think?

i think that it could be a good practice to add to our provisioner (specially once we will have the unified provisioner in place), but out of scope for this task.
@soyacz , please open a task for it (in the New column), and it can address much more than only AZ, but also region, based on price and availability.

**Comment by fgelcer** [comment]
> > re if this solution works:
> > ```
> > 1. you tested with artifact test which is not using provisioner used in separate provision-step used in longevities. Artifact test is not using that part of code.
> > 
> > 2. How it will behave when we have sct-runner in one AZ while resources in another? We should also connect runner to resources subnet as well i think
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > besides that, I did some investigation about solving our problem and found that AWS have an api to recommend best region/AZ for given requested resources: https://docs.aws.amazon.com/cli/latest/reference/ec2/get-spot-placement-scores.html
> > This could be also solution to this problem and could be much cleaner (e.g. based on test resources, before creating sct runner, get best AZ (or even region) for test and create runner and resources there.)
> 
> @soyacz ,
> 
> 1. in principle and for the long term it sounds correct. but as you can note, the goal for this task is only a minimal fix to mainly solve the problem we face in artifacts test. So it is not involved with any refactor, just a patch.
> 2. artifacts test doesn't use sct runner.
> 3. This fix was tested and passed ok for artifacts test.
@yarongilor did you test also some other test, that does use SCT Runner to be sure it won't break anything else?

> 4. I'm not sure it is a problem if sct runner located in a different zone.

**Comment by roydahan** [comment]
No need to open a new task, we will open one if we face any issue.
Here we wanted to solve the issue of artifacts failing.
Solution for something else isn’t required.

**Comment by fgelcer** [comment]
> No need to open a new task, we will open one if we face any issue. Here we wanted to solve the issue of artifacts failing. Solution for something else isn’t required.

too late, there is already an improvement task open:
https://github.com/scylladb/scylla-cluster-tests/issues/5201

**Comment by yarongilor** [comment]
@fgelcer , tested ok in another (non artifacts) test in:
https://jenkins.scylladb.com/job/scylla-staging/job/Longevity_yaron/job/longevity-large-partition-8h-test-rq/3/

Screenshot from 2022-08-30 10-44-48

**Comment by soyacz** [comment]
> @fgelcer , tested ok in another (non artifacts) test in: https://jenkins.scylladb.com/job/scylla-staging/job/Longevity_yaron/job/longevity-large-partition-8h-test-rq/3/
> 
> Screenshot from 2022-08-30 10-44-48

Yes, looks it works - you may stop resources - no need to run full test.
